torch/distributed/fsdp/flat_param.py

```diff
@@ -312,87 +368,43 @@ def flatten_params(
         """
-        Unflattens the unsharded flattened parameter by setting the original
-        module parameter variables to be views into it.
-
-        Args:
-            as_params (bool): If ``True``, then registers the original
-                parameters as ``nn.Parameter`` s; if ``False``, then registers
-                the original parameters only as ``Tensor`` s. ``False`` should
-                be used during forward/backward computation and when hiding the
-                original parameters from :meth:`nn.Module.named_parameters`.
-        """
-        views = self._get_unflat_views(self.flat_param)
-        for view, (param_name, module, _) in zip(views, self.flat_param._param_infos):
-            if hasattr(module, param_name):
-                delattr(module, param_name)
-            if as_params:
-                module.register_parameter(param_name, nn.Parameter(view))
-            else:
-                setattr(module, param_name, view)
-        for (
-            param_name,
-            module,
-            _,
-            prim_param_name,
-            prim_module,
-            _,
-        ) in self.flat_param._shared_param_infos:
-            if hasattr(module, param_name):
-                delattr(module, param_name)
-            assert hasattr(prim_module, prim_param_name)
-            param: Union[Tensor, nn.Parameter] = getattr(prim_module, prim_param_name)
-            if as_params:
-                assert isinstance(param, nn.Parameter)
-                module.register_parameter(param_name, param)
-            else:
-                setattr(module, param_name, param)
-
-    @contextlib.contextmanager
-    def unflatten_as_params(self) -> Generator:
-        """
-        Assumes the flattened parameter is unsharded. When in the context,
-        unflattens the original parameters as ``nn.Parameter`` views into the
-        flattened parameter, and after the context, restores the original
-        parameters as ``Tensor`` views into the flattened parameter.
-        """
-        self._unflatten(as_params=True)
-        try:
-            yield
-        finally:
-            self._unflatten(as_params=False)
+        if not self.uses_sharded_strategy:
+            return
+        flat_param = self.flat_param
+        self.process_group = process_group
+        self.rank = process_group.rank()
+        self.world_size = process_group.size()
+        assert (
+            flat_param.storage_offset() == 0
+        ), "The `FlatParameter` is not the sole occupant of its storage"
+        orig_storage = flat_param.storage()
+        local_shard, numel_padded = FlatParamHandle._get_shard(
+            flat_param, self.rank, self.world_size
+        )
+        flat_param.set_(local_shard)  # type: ignore[call-overload]
+        self._init_shard_metadata(local_shard.numel(), numel_padded, self.rank)
+        if orig_storage.size() > 0:
+            orig_storage.resize_(0)
```

Comment:
call `_free_storage()`?

Reply:
Ah. The reason we do not call `_free_storage()` is because `_free_storage()` takes in a `Tensor`, not a `Storage`. I am pretty sure that we do not even need to save the variable `orig_storage` and call `orig_storage.resize_(0)` explicitly. After calling `flat_param.set_(local_shard)`, the original storage should have a reference count of 0 and be naturally garbage collected. For this PR, I will leave it like this.